velox/exec/TopNRowNumber.cpp

```diff
@@ -194,13 +203,35 @@ void TopNRowNumber::addInput(RowVectorPtr input) {
       auto& partition = partitionAt(lookup_->hits[i]);
       processInputRow(i, partition);
     }
+
+    if (abandonPartialEarly()) {
+      abandonedPartial_ = true;
+      addRuntimeStat("abandonedPartial", RuntimeCounter(1));
+
+      updateEstimatedOutputRowSize();
+      outputBatchSize_ = outputBatchRows(estimatedOutputRowSize_);
+      outputRows_.resize(outputBatchSize_);
+    }
   } else {
     for (auto i = 0; i < numInput; ++i) {
       processInputRow(i, *singlePartition_);
     }
   }
 }
 
+bool TopNRowNumber::abandonPartialEarly() const {
+  if (table_ == nullptr || generateRowNumber_ || spiller_ != nullptr) {
+    return false;
+  }
+
+  const auto numInput = stats_.rlock()->inputPositions;
+  const auto numOutput = data_->numRows() +
+      (spiller_ == nullptr ? 0 : spiller_->stats().spilledRows);
+
+  return numInput > abandonPartialMinRows_ &&
```

Comment:
Shall we check "numInput > abandonPartialMinRows_" early?

Reply:
Sure. Will do.